microsoft/testsuites/sysbench/sysbenchperf.py

```diff
@@ -0,0 +1,406 @@
+    )
+    def verify_sysbench_memory_perf(
+        self,
+        log: Logger,
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        memory_operation: List[str] = ["read", "write"]
+        memory_access_mode: List[str] = ["seq", "rnd"]
+        failure = []
+        sysbench = node.tools[Sysbench]
+
+        for op in memory_operation:
+            for access_mode in memory_access_mode:
+                subtest_name: str = f"{access_mode}_{op}_default_memory_perf_check"
+                res: SysbenchTestResult = SysbenchTestResult()
+                output: ExecutableResult = sysbench.run_memory_perf(
+                    memory_access_mode=access_mode,
+                    memory_oper=op,
+                )
+
+                if output.exit_code != 0:
+                    failure.append(f"{output.cmd}")
+                    res.name = subtest_name
+                    res.message = (
+                        f"{subtest_name} failed with " f"exit code {output.exit_code}"
+                    )
+                    res.status = TestStatus.FAILED
+                else:
+                    res = process_memory_perf_result(
+                        output.stdout,
+                        subtest_name,
+                    )
+
+                self._send_subtest_msg(
+                    result,
+                    res.name,
+                    res.status,
+                    res.message,
+                )
+
+        if len(failure) > 0:
+            fail(f"Sysbench-memory test failed: {failure}")
+
+    def _send_subtest_msg(
+        self,
+        test_result: TestResult,
+        test_name: str,
+        test_status: TestStatus,
+        test_message: str = "",
+    ) -> None:
+        subtest_msg = send_sub_test_result_message(
+            test_result=test_result,
+            test_case_name=test_name,
+            test_status=test_status,
+            test_message=test_message,
+        )
+
+        notifier.notify(subtest_msg)
+
+
+def process_perf_result(data: str, testcase_name: str) -> SysbenchTestResult:
+    non_debug_pattern = r"^(?!.*DEBUG: ).*$"
+    debug_patern = r".*DEBUG: .*"
+
+    result: Dict[str, Any] = {}
+
+    # Find all non-DEBUG lines in the text
+    non_debug_text = "\n".join(
+        re.findall(
+            non_debug_pattern,
+            data,
+            re.MULTILINE,
+        )
+    )
+    debug_text = "\n".join(
+        re.findall(
+            debug_patern,
+            data,
+            re.MULTILINE,
+        )
+    )
+
+    # Extract total time using regular expression
+    total_time = re.search(
+        r"total time:\s+([\d.]+s)",
+        non_debug_text,
+    )
+    if total_time:
+        result["total_time"] = total_time.group(1)
+
+    # Extract total number of events using regular expression
+    total_events = re.search(
+        r"total number of events:\s+(\d+)",
```

Comment:
1. Add examples on each regexp pattern.
2. Move them out of the function, and group them together, so it's easy to maintain together.
3. Use named captures to improve readability and easier to maintain.
4. Use methods in utils/__init__.py, like `find_group_in_lines`, `find_groups_in_lines`.

Reply:
I added regex examples for reference and using named captures now.

`find_group_in_lines `is not giving me correct result. I have tried below thing for `total number of events`

I passed `single_lines=False` which gives me `{'total_events': '2'}` whereas correct result is  `'total_events': '23274'`

I also printed `find ` variable value in `find_patterns_groups_in_lines `function which seems correct. Issue seems at `results[index].append(dict(zip(pattern.groupindex, find)))`

```
find_group_in_lines(
            lines=data,
            pattern=total_events_pattern,
            single_line=False,
        )
```

Reply:
You can add `$` at the end to match all numbers in last pattern. The (\d+) should match all numbers in greedy mode, but I don't know why it doesn't work.

Reply:
Yes i Tried that too but same result.

Reply:
Good catch! I will fix it. Before that, please remove the single_line=False. It doesn't need for this pattern.

Reply:
Yes, i tried without the single_line=False and result was empty. I am not using find_group_in_lines in the code as of now.

Reply:
I merged the fix, please use `find_group_in_lines` or similar in all places.

Reply:
Done